File: src/test/java/idlezoo/security/SecurityTest.java
package idlezoo.security;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.http.*;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;

import static java.util.Collections.singletonList;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(SpringExtension.class)
@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
class SecurityTest {
    private static final String ORIGIN = "http://localhost:9000";

    private final PasswordEncoder passwordEncoder;
    private final TestRestTemplate template;
    private final int port;

    @MockBean
    private UsersService usersService;

    @Autowired
    SecurityTest(PasswordEncoder passwordEncoder, @LocalServerPort int port, TestRestTemplate template) {
        this.passwordEncoder = passwordEncoder;
        this.port = port;
        this.template = template;
    }

    @Test
    void homePageRedirects() {
        ResponseEntity<String> response = template.getForEntity("/", String.class);
        assertEquals(HttpStatus.FOUND, response.getStatusCode());
        assertEquals("https://idlezoo.github.io", response.getHeaders().getFirst(HttpHeaders.LOCATION));
    }

    @Test
    void userEndpoint() {
        ResponseEntity<String> response = template.getForEntity("/user", String.class);
        assertEquals(HttpStatus.OK, response.getStatusCode());
        assertNull(response.getBody());
    }

    @Test
    void cors() throws Exception {
        user();
    }

    private <T> void assertCors(ResponseEntity<T> response) {
        assertTrue(response.getHeaders().getAccessControlAllowCredentials());
        assertEquals(ORIGIN, response.getHeaders().getAccessControlAllowOrigin());
    }

    @Test
    void register() throws Exception {
        loginOrRegister("testregister", "/createuser");
        verify(usersService).addUser(eq("testregister"), anyString());
    }

    @Test
    void login() throws Exception {
        when(usersService.loadUserByUsername("testuser"))
                .thenReturn(
                        new IdUser(0, "testuser", passwordEncoder.encode("testuser"))
                );
        ResponseEntity<Void> login = loginOrRegister("testuser", "/login");
        assertEquals(HttpStatus.OK, login.getStatusCode());
        List<String> cookies = login.getHeaders().get(HttpHeaders.SET_COOKIE);
        assertNotNull(cookies);
        ResponseEntity<String> user =
                user(cookies.toArray(new String[0]));

        assertNotNull(user.getBody());
        assertTrue(user.getBody().contains("testuser"));
    }

    private ResponseEntity<String> user(String... cookies) throws URISyntaxException {
        RequestEntity<Void> request =
                RequestEntity.get(new URI("http://localhost:" + port + "/user"))
                        .header(HttpHeaders.ORIGIN, ORIGIN)
                        .header(HttpHeaders.COOKIE, cookies)
                        .build();

        ResponseEntity<String> response = template.exchange(request, String.class);
        assertEquals(HttpStatus.OK, response.getStatusCode());
        assertCors(response);
        return response;
    }

    private ResponseEntity<Void> loginOrRegister(String user, String uri) throws URISyntaxException {
        RequestEntity<Void> request =
                RequestEntity.get(new URI("http://localhost:" + port + "/resource"))
                        .header(HttpHeaders.ORIGIN, ORIGIN)
                        .build();

        ResponseEntity<String> response = template.exchange(request, String.class);
        assertCors(response);

        MultiValueMap<String, String> form = new LinkedMultiValueMap<>();
        form.set("username", user);
        form.set("password", user);
        HttpHeaders headers = new HttpHeaders();

        headers.put(HttpHeaders.ORIGIN, singletonList(ORIGIN));
        RequestEntity<MultiValueMap<String, String>> loginOrRegister = new RequestEntity<>(
                form, headers, HttpMethod.POST, URI.create("http://localhost:" + port + uri));
        ResponseEntity<Void> result = template.exchange(loginOrRegister, Void.class);
        assertCors(result);
        return result;
    }

}

File: src/test/kotlin/idlezoo/security/SecurityTest.kt
package idlezoo.security

import com.nhaarman.mockito_kotlin.whenever
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito.verify
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.*
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.util.LinkedMultiValueMap
import org.springframework.util.MultiValueMap
import java.net.URI

@ExtendWith(SpringExtension::class)
@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
class SecurityTest(@Autowired val passwordEncoder: PasswordEncoder,
                   @Autowired val template: TestRestTemplate
) {

    @MockBean
    private lateinit var usersService: UsersService

    @Test
    fun homePageRedirects() {
        val response = template.getForEntity("/", String::class.java)
        assertEquals(HttpStatus.FOUND, response.statusCode)
        assertEquals("https://idlezoo.github.io", response.headers.getFirst(HttpHeaders.LOCATION))
    }

    @Test
    fun userEndpoint() {
        val response = template.getForEntity("/user", String::class.java)
        assertEquals(HttpStatus.OK, response.statusCode)
        assertNull(response.body)
    }

    @Test
    fun cors() {
        user()
    }

    private fun <T> assertCors(response: ResponseEntity<T>) {
        assertTrue(response.headers.accessControlAllowCredentials)
        assertEquals(ORIGIN, response.headers.accessControlAllowOrigin)
    }

    @Test
    fun register() {
        loginOrRegister("testregister", "/createuser")
        verify(usersService).addUser(eq("testregister"), anyString())
    }

    @Test
    fun login() {
        whenever(usersService.loadUserByUsername("testuser"))
                .thenReturn(
                        IdUser(0, "testuser", passwordEncoder.encode("testuser"))
                )
        val login = loginOrRegister("testuser", "/login")
        assertEquals(HttpStatus.OK, login.statusCode)
        val cookies = login.headers[HttpHeaders.SET_COOKIE]
        assertNotNull(cookies)
        val user = user(*cookies!!.toTypedArray())

        assertNotNull(user.body)
        assertTrue(user.body!!.contains("testuser"))
    }

    private fun user(vararg cookies: String): ResponseEntity<String> {
        val request = RequestEntity.get(URI("/user"))
                .header(HttpHeaders.ORIGIN, ORIGIN)
                .header(HttpHeaders.COOKIE, *cookies)
                .build()

        val response = template.exchange(request, String::class.java)
        assertEquals(HttpStatus.OK, response.statusCode)
        assertCors(response)
        return response
    }

    private fun loginOrRegister(user: String, uri: String): ResponseEntity<Void> {
        val request = RequestEntity.get(URI("/resource"))
                .header(HttpHeaders.ORIGIN, ORIGIN)
                .build()

        val response = template.exchange(request, String::class.java)
        assertCors(response)

        val form = LinkedMultiValueMap<String, String>()
        form.set("username", user)
        form.set("password", user)
        val headers = HttpHeaders()

        headers[HttpHeaders.ORIGIN] = listOf(ORIGIN)
        val loginOrRegister = RequestEntity<MultiValueMap<String, String>>(form, headers, HttpMethod.POST, URI(uri))
        val result = template.exchange(loginOrRegister, Void::class.java)
        assertCors(result)
        return result
    }

    companion object {
        private val ORIGIN = "http://localhost:9000"
    }

}
